pkg/reconciler/build/build.go

```diff
@@ -137,63 +148,76 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	// Don't mutate the informer's copy of our object.
 	build = build.DeepCopy()
 
-	// Validate build
-	if err = c.validateBuild(build); err != nil {
-		c.Logger.Errorf("Failed to validate build: %v", err)
-		return err
-	}
-
 	// If the build's done, then ignore it.
-	if builder.IsDone(&build.Status) {
+	if isDone(&build.Status) {
 		return nil
 	}
 
-	// If the build is not done, but is in progress (has an operation), then asynchronously wait for it.
-	// TODO(mattmoor): Check whether the Builder matches the kind of our c.builder.
-	if build.Status.Builder != "" {
-		op, err := c.builder.OperationFromStatus(&build.Status)
-		if err != nil {
+	// If the build hasn't started yet, validate it and create a Pod for it
+	// and record that pod's name in the build status.
+	var p *corev1.Pod
+	if build.Status.Cluster == nil || build.Status.Cluster.PodName == "" {
+		if err = c.validateBuild(build); err != nil {
+			logger.Errorf("Failed to validate build: %v", err)
 			return err
 		}
-
-		// Check if build has timed out
-		if builder.IsTimeout(&build.Status, build.Spec.Timeout) {
-			//cleanup operation and update status
-			timeoutMsg := fmt.Sprintf("Build %q failed to finish within %q", build.Name, build.Spec.Timeout.Duration.String())
-
-			if err := op.Terminate(); err != nil {
-				c.Logger.Errorf("Failed to terminate pod: %v", err)
-				return err
-			}
-
+		p, err = c.startPodForBuild(build)
+		if err != nil {
 			build.Status.SetCondition(&duckv1alpha1.Condition{
 				Type:    v1alpha1.BuildSucceeded,
 				Status:  corev1.ConditionFalse,
-				Reason:  "BuildTimeout",
-				Message: timeoutMsg,
+				Reason:  "BuildExecuteFailed",
+				Message: err.Error(),
 			})
-			// update build completed time
-			build.Status.CompletionTime = &metav1.Time{time.Now()}
-
-			if _, err := c.updateStatus(build); err != nil {
-				c.Logger.Errorf("Failed to update status for pod: %v", err)
+			if err := c.updateStatus(build); err != nil {
 				return err
 			}
-
-			c.Logger.Errorf("Timeout: %v", timeoutMsg)
-			return nil
+			return err
 		}
+	} else {
+		// If the build is ongoing, update its status based on its pod, and
+		// check if it's timed out.
+		p, err = c.podsLister.Pods(build.Namespace).Get(build.Status.Cluster.PodName)
+		if err != nil {
+			// TODO: What if the pod is deleted out from under us?
+			return err
+		}
+	}
 
-		// if not timed out then wait async
-		go c.waitForOperation(build, op)
-		return nil
+	// Update the build's status based on the pod's status.
+	build.Status = resources.BuildStatusFromPod(p, build.Spec)
+
+	// Check if build has timed out; if it is, this will set its status
+	// accordingly.
+	if err := c.checkTimeout(build); err != nil {
+		return err
+	}
+
+	return c.updateStatus(build)
+}
+
+func (c *Reconciler) updateStatus(u *v1alpha1.Build) error {
+	newb, err := c.buildclientset.BuildV1alpha1().Builds(u.Namespace).Get(u.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
 	}
+	newb.Status = u.Status
```

Comment:
We could consider checking if there is any change in status object before updating it.
This could be changed in further PR too.